### PR TITLE
Feature/add anmm

### PR DESCRIPTION
**Pull request by ChrisRBXiong**
#61 
add aNMM model

### CODE REVIEW

File: matchzoo/models/anmm.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
+"""An implementation of aNMM Model."""
+import math
+import typing
+
+import numpy as np
+import torch
+import torch.nn as nn
```
Comment (caiyinqiong) [inline matchzoo/models/anmm.py]: Add a blank line.

File: matchzoo/models/anmm.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
+"""An implementation of aNMM Model."""
+import math
+import typing
+
+import numpy as np
```
Comment (Chriskuei) [inline matchzoo/models/anmm.py]: unused package

File: matchzoo/models/anmm.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+            )
+            for in_size, out_size, in zip(
+                in_hidden_size,
+                out_hidden_size
+            )
+        ]
+        self.hidden_layers = nn.Sequential(*hidden_layers)
+
+        # Query Attention
+        self.q_attention = nn.Linear(self._params['embedding_output_dim'], 1, bias=False)
```
Comment (Chriskuei) [inline matchzoo/models/anmm.py]: Use `Attention` module to rewrite. And remove `math` package.
https://github.com/NTMC-Community/MatchZoo-py/blob/f4c28583b7cd076dd128ea0916ed49a372b98f14/matchzoo/modules/attention.py#L9-L37

File: matchzoo/models/anmm.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        # apply dropout
+        bin_qa_matching = self.dropout(bin_qa_matching)
+
+        # MLP hidden layers
+        # shape = [B, L, 1]
+        hiddens = self.hidden_layers(bin_qa_matching)
+
+        # query attention
+        # shape = [B, L, 1]
+        q_attention = self.q_attention(embed_left).unsqueeze(-1)
```
Comment (caiyinqiong) [inline matchzoo/models/anmm.py]: The inputs of `Attention` module are left and left_mask.

File: matchzoo/models/anmm.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        # shape = [B, R]
+        input_left, input_right = inputs['text_left'], inputs['text_right']
+
+        # Process left and right input.
+        # shape = [B, L, D]
+        # shape = [B, R, D]
+        embed_left = self.embedding(input_left.long())
+        embed_right = self.embedding(input_right.long())
+
+        # right input mask matrix
```
Comment (caiyinqiong) [inline matchzoo/models/anmm.py]: `Right input mask matrix`

File: matchzoo/models/anmm.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+                                    device=device).long().unsqueeze(-1)
+        bin_indexes += index_offset
+        bin_indexes = bin_indexes.view(-1)
+
+        # shape = [B, L, BI]
+        bin_qa_matching = torch.zeros(B * L * BI, device=device)
+        bin_qa_matching.index_add_(0, bin_indexes, qa_matching_matrix)
+        bin_qa_matching = bin_qa_matching.view(B, L, -1)
+
+        # apply dropout
```
Comment (caiyinqiong) [inline matchzoo/models/anmm.py]: Apply dropout

File: matchzoo/models/anmm.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+        bin_qa_matching = self.dropout(bin_qa_matching)
+
+        # MLP hidden layers
+        # shape = [B, L, 1]
+        hiddens = self.hidden_layers(bin_qa_matching)
+
+        # query attention
```
Comment (caiyinqiong) [inline matchzoo/models/anmm.py]: Query attention
